src/dso_api/dynamic_api/filters/openapi.py

```diff
@@ -0,0 +1,177 @@
+    as that would explode in a huge combination.
+    But it will suffice for most typical users.
+    """
+    openapi_params = []
+
+    for field in table_schema.get_fields(include_subfields=False):
+        if field.is_array_of_objects:
+            # M2M relation or nested table.
+            for sub_field in field.subfields:
+                openapi_params.extend(_get_field_openapi_params(sub_field))
+        else:
+            # Flat direct field
+            openapi_params.extend(_get_field_openapi_params(field))
+
+    return openapi_params
+
+
+def _get_field_openapi_params(field: DatasetFieldSchema, prefix="") -> list[dict]:  # noqa: C901
+    """Generate the possible OpenAPI filter parameters for a single field."""
+    if field.parent_field:
+        prefix = f"{prefix}{field.parent_field.name}."
+
+    openapi_params = []
+    for lookup in QueryFilterEngine.get_allowed_lookups(field):
+        param = {
+            "name": _get_filter_name(prefix, field, lookup),
+            "in": "query",
+            "description": _get_filter_description(field, lookup=lookup),
+            "schema": {},
+        }
+
+        if field.is_geo:
+            param["schema"] = {"type": "string"}  # POINT(x y) or "x,y"
+        elif type := OPENAPI_LOOKUP_TYPES.get(lookup):
+            # Ignore field type, lookup has different type.
+            param["schema"] = {"type": str(type)}
+        else:
+            # Use standard OpenAPI type/format like our JSONSchema also does.
+            param["schema"] = {"type": field.type}
+            if field.format:
+                param["schema"]["format"] = field.format
+
+            if field.is_array:
+                param["schema"]["items"] = field["items"]  # expose all, is both JSONSchema.
+
+            if lookup in OPENAPI_COMMA_LOOKUPS:
+                # OpenAPI has a specific notation to encode "val1,val2,.." formats.
+                # See: https://swagger.io/docs/specification/serialization/#query
+                param["explode"] = False
+                param["style"] = "form"
+                param["schema"] = {"type": "array", "items": param["schema"]}
+
+        if lookup in parser.MULTI_VALUE_LOOKUPS:
+            # The 'field[not]=..' parameter can be repeated.
+            param["schema"] = {
+                "type": "array",
+                "items": param["schema"],
+            }
+
+        openapi_params.append(param)
+
+    return openapi_params
+
+
+def _get_filter_name(prefix: str, field: DatasetFieldSchema, lookup: str) -> str:
+    """Generate the filter name.
+    This generates the "relation.field[lookup]" notation.
+    """
+    name = f"{prefix}{field.name}"
+    if field.relation and not field.is_loose_relation:
+        name += "Id"  # ForeignKey field
+    if lookup:
+        name = f"{name}[{lookup}]"
+    return name
+
+
+def _get_filter_description(field: DatasetFieldSchema, lookup=None) -> str | dict:  # noqa: C901
```

Comment:
pipe notation for types is python3.10 only. Is mypy borking on this?

Reply:
nope! 🤩 because the file has `from __future__ import annotations` in it! :-D